mono/utils/mono-threads.h

```diff
@@ -153,6 +153,8 @@ enum {
 	ASYNC_SUSPEND_STATE_INDEX = 1,
 };
 
+struct HandleStack;
+struct MonoJitTlsData;
```

Comment:
should we follow the underscore convention for these struct tags?

Reply:
Maybe, maybe not.
Eventually you just have C++ "struct foo" and the underscores kinda get in the way -- they end up duplicated.
For C it is a good convention, it implies private -- users should say `Foo` not `struct _Foo`.

Reply:
bikeshedding++:


I suggest the ugliness of the eventual hypothetical C++ outweighs the C convention, even if the C++ does not come to pass. But it is arguable either way. You could remove the underscores later. You could leave them in the C++. The names are also reserved for the C/C++ implementer but *everyone* ignores that part and gets away with it.
The underscores are part of the C++ ABI, the name mangling, but that doesn't clearly  matter either. You can strive for an externC ABI, or you can break the C++ ABI. The intersection of C++ ABI and compatibility is approximately zero. glibc has broken those sorts of things.

I believe Mono follows both conventions, which isn't to say anyone agreed to it.

Anyway, independent of the leading underscore, I believe this is a good coding style. It has cons but mostly pros:

pros:
  can see things w/o cast in debugger
  can pass around  the pointers among clients w/o cast; can stop writing the little wrapper macros and functions to do the  casting (!) (which arguably takes a popular breakpoint target away)
  The struct is still usually totally opaque to client, as it is not declared where they see it, typically. So just as good as a gpointer.

cons:
  Revealing something is a struct is a certain limited transparency. You are revealing it isn't a  union or an integer. But yet, really, it could be. The struct is just a  "token" type, and internally  you can cast. Albeit, you are limited then to the size. If you wanted 64bits on 32bit platform, the struct* isn't good  -- in that case, pass a struct by value that has a longlong, or however many  opaque bits you need. gpointer has same problem.

   There is still no provision for a hiearchy of  types.  Imagine we had object, string, and animal. They'd each be separate uncovertable types. If they were all opaque integer or void*, they'd be freely convertable, perhaps too freely.

 That it is a struct foo* is a cat somewhat out of the bag. You then have to be careful to guard the struct declaration and prevent its accidental misuse. A gpointer is harder to accidentally abuse as it requires a cast. The  struct might be revealed by accident as it is often just an include away. Though in a really nicely formed system, the struct is only revealed in a .c file, not any foo-internals.h. That is good. Imagine threads-types.h only had extra functions, and the types were only  in threads.c